mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java

```diff
@@ -91,16 +105,22 @@ void start() {
     }
 
     private Mono<Void> doDispatch(Mono<byte[]> serializedEvent, Set<RegistrationKey> keys) {
-        Flux<RoutingKeyConverter.RoutingKey> routingKeys = Flux.concat(
-            Mono.just(RoutingKeyConverter.RoutingKey.empty()),
-            Flux.fromIterable(keys)
-                .map(RoutingKeyConverter.RoutingKey::of));
+        Stream<RoutingKeyConverter.RoutingKey> routingKeyStream = Stream.concat(
+            Stream.of(RoutingKeyConverter.RoutingKey.empty()),
+            keys.stream().map(RoutingKeyConverter.RoutingKey::of));
+
+        return doDispatch(serializedEvent, MAILBOX_EVENT_EXCHANGE_NAME, routingKeyStream, basicProperties);
+    }
 
-        Flux<OutboundMessage> outboundMessages = routingKeys
-            .flatMap(routingKey -> serializedEvent
-                .map(payload -> new OutboundMessage(MAILBOX_EVENT_EXCHANGE_NAME, routingKey.asString(), basicProperties, payload)));
+    Mono<Void> doDispatch(Mono<byte[]> serializedEvent, String exchangeName, Stream<RoutingKeyConverter.RoutingKey> routingKeyStream, AMQP.BasicProperties properties) {
+        Stream<OutboundMessage> outboundMessages = routingKeyStream
+            .map(routingKey -> new OutboundMessage(exchangeName, routingKey.asString(), properties, serializedEvent.block()));
 
-        return sender.send(outboundMessages);
+        return Mono.fromRunnable(() -> {
+            synchronized (lock) {
```

Comment:
why did you need to lock? I think any reactor object is suppose to work well under concurrency environtment

Reply:
`I think any reactor object is suppose to work well under concurrency environtment`

This assumption is wrong: the producer needs to be on a single thread to cope with Rx specifications.

Please have a look to FluxProcessor subclass javadoc to get a glance of the difficulty of that topic.

Reply:
A lock ensures things go fine.

Reply:
what is different if I created a method likes below and use it in Mono.fromRunnable()? I think the code will look better. Anw, that is just a tiny comment.
```java
sychronized emit(OutboundMessage message) {
    emmiterProcessor.onNext(message);
}
```